serverless/appsec: ASM standalone mode

Hellzy opened the pull request:
### What does this PR do?

This PR is part of the Standalone Billing ASM initiative for aws-lambda, which aims at enabling ASM as a (seemingly) standalone product from a customer billing perspective. You can find all details in the following [RFC](https://docs.google.com/document/d/1BGfQqVQn0TK-JYQnM-Rn3PJrSHOGeEbky91PGbFFZX0/edit?pli=1#heading=h.hs091bhdmugz).

- Add a new env var `DD_APM_TRACING_ENABLED`. The new ASM standalone state can then be determined
by checking that this new env var is set to `false` and that ASM is enabled
- Modify the trace agent priority sampling rate to approach 1 trace/min in case of standalone mode. This results in two cases:
    - In case of an attack, the trace is set with manual keep priority, overriding sampling limitations
    - In any other case, that trace is sampled with a 1/min rate, which is the decided limit for ASM's integration
       with span tracing to still function, while suppressing APM data.
- Set the tag `_dd.apm.enabled` to 0 at the chunk level in ASM standalone mode (currently also done at the span level while we wait for backend to handle the chunk level tag)
- Log a message when starting in ASM standalone mode
      

### Motivation

These changes are needed to comply with the RFC for ASM standalone billing. ASM want's to implement standalone billing for customers to be allowed to only be billed for ASM products as opposed to now where they'll be billed for other products as well (ex: APM). While implementing a fully standalone product is possible, it would require a lot of redesign and take a considerable amount of time, as opposed to the chosen solution which instead chooses to mask trace information that isn't relevant to ASM.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

Hellzy committed:
serverless/appsec: change sampling and tags in case of standalone ASM

Julio-Guerra commented on pkg/serverless/appsec/httpsec/tags.go:
```diff
@@ -78,6 +78,11 @@ type span interface {
+	value := os.Getenv("DD_APM_TRACING_ENABLED")
+	standalone := value != "true" && value != "1"
+	if standalone {
+		span.SetMetricsTag("_dd.apm.enabled", 0)
```
Must be set in the trace chunk tags now.

Hellzy replied:
Done in e8da65f67931ed9b

Hellzy committed:
pkg/serverless/appsec: set _dd.apm.enabled at chunk level

Hellzy committed:
cmd/serverless: explain ASM rate limiting override

Hellzy committed:
serverless: appsec: adjust sampling and tag every span for standalone

Hellzy committed:
appsec: standalone: tag spans unconditionally && rework isStandalone check

RomainMuller commented on cmd/serverless/main.go:
```diff
@@ -326,6 +326,11 @@ func runAgent(stopCh chan struct{}) (serverlessDaemon *daemon.Daemon, err error)
+		// Set the default rate limiting to 1 trace/min to limit non ASM related traces as much as possible.
+		// In case of ASM event, the trace priority will be set to manual keep
+		if appsec.IsStandalone() {
+			ta.PrioritySampler.UpdateTargetTPS(1. / 120)
```
I find it somewhat confusing that "1 per minute" is `1/120`? Can this be explained (at least to me, if that is clear to anyone else)?

Also - this looks like a magic number, which might be best shoved into a constant.

Julio-Guerra replied:
It's an overall APM platform limitation based on constant streaming of traces and spans. Making it 0, which would be "send nothing but traces with appsec events", would break the billing pipeline and service catalog, which both expect receive traces. And we identified that the minimum we could get is 1 trace per minute.

Julio-Guerra replied:
Worth further detailing this in the code comment indeed @Hellzy 🙏

Hellzy replied:
Enhanced the comment. As for the number, it indeed reflects 1/2min, but it's interestingly the threshold that reflects 1/min when continuously testing under various loads

Hellzy replied:
721cd746ac13e

RomainMuller commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -236,3 +253,9 @@ type bytesStringer []byte
 	return string(b)
 }
+
+func isStandalone() bool {
+	value, set := os.LookupEnv("DD_APM_TRACING_ENABLED")
+	enabled, _ := strconv.ParseBool(value)
+	return set && !enabled
+}
```
There is an identical exported function in `serverless/appsec`, it should be used instead of duplicated.

Hellzy replied:
Refactored it in the new `config` pkg, now used in `cmd/serverless` and internally. This is to avoid circular dependencies.
See 25c20caf62ffdde for the new test.

RomainMuller commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -219,10 +230,16 @@ func (lp *ProxyLifecycleProcessor) WrapSpanModifier(ctx ExecutionContext, modify
 }
 
 func (s *spanWrapper) SetMetricsTag(tag string, value float64) {
+	if s.Metrics == nil {
+		s.Metrics = make(map[string]float64)
+	}
```
Why was this not necessary before?

RomainMuller commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -219,10 +230,16 @@ func (lp *ProxyLifecycleProcessor) WrapSpanModifier(ctx ExecutionContext, modify
 type spanWrapper pb.Span
 
 func (s *spanWrapper) SetMetaTag(tag string, value string) {
+	if s.Meta == nil {
+		s.Meta = make(map[string]string)
+	}
```
Why was this not necessary before?

IvanTopolcic reviewed (approved):
Looks good from my end, I think mostly just removing the duplicated code would be 👍

Julio-Guerra commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -95,7 +97,16 @@ func (lp *ProxyLifecycleProcessor) OnInvokeEnd(_ *invocationlifecycle.Invocation
+			chunk.Tags = make(map[string]string)
+		}
+		chunk.Tags["_dd.apm.enabled"] = "0"
+		for _, s := range chunk.Spans {
```
Worth adding a comment saying that the specification says service entry span only but we are putting them everywhere here for extra care (in case the top-level span condition isn't properly met at this point of the agent (because the agent can automatically change a span into a service entry span at some point if the service name is different or if there is no parent span, but I don't know where compared to this function call)).

Julio-Guerra reviewed (changes requested):
Thank you, I left some comments

Julio-Guerra commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -95,7 +97,16 @@ func (lp *ProxyLifecycleProcessor) OnInvokeEnd(_ *invocationlifecycle.Invocation
-	// Add appsec tags to the aws lambda function service entry span
+	// Add relevant standalone tags to the chunk (TODO: remove per span tagging once backend handles chunk tags)
+	if isStandalone() {
+		if chunk.Tags == nil {
+			chunk.Tags = make(map[string]string)
+		}
+		chunk.Tags["_dd.apm.enabled"] = "0"
```
Sorry, false alarm.

Hellzy committed:
Review: enhance comments

Hellzy committed:
serverless/appsec: move appsec config to "config" pkg

Having a config package allows to use some config functions both
internally and externally. A prime usefull example is using
config.IsStandalone() both in the http proxy and in cmd/serverless

Hellzy committed:
serverless/appsec: add config.IsStandalone() test

Julio-Guerra commented on pkg/serverless/appsec/config/config.go:
```diff
@@ -70,16 +71,24 @@ func isEnabled() (enabled bool, set bool, err error) {
 
-func newConfig() (*Config, error) {
+// IsStandalone returns whether appsec is used as a standalone product (without APM tracing) or not
+func IsStandalone() bool {
+	value, set := os.LookupEnv(tracingEnabledEnvVar)
+	enabled, _ := strconv.ParseBool(value)
+	return set && !enabled
+}
```
This function call is costly so I suggest you cache the result to avoid extra latencies in the request hot path where you use it.

Hellzy replied:
6050d529532bd634

RomainMuller reviewed (approved):
LGTM granted the last comments from @Julio-Guerra are taken care of, so I'm approving here so as not to block it anymore.

Julio-Guerra commented on pkg/serverless/appsec/httpsec/proxy.go:
```diff
@@ -99,7 +100,18 @@ func (lp *ProxyLifecycleProcessor) OnInvokeEnd(_ *invocationlifecycle.Invocation
 func (lp *ProxyLifecycleProcessor) spanModifier(lastReqId string, chunk *pb.TraceChunk, s *pb.Span) {
-	// Add appsec tags to the aws lambda function service entry span
+	// Add relevant standalone tags to the chunk (TODO: remove per span tagging once backend handles chunk tags)
+	if config.IsStandalone() {
```
Even if you optimize `IsStandalone()` to make it efficient for this call, you could also store at init-time the global span tags in `lp` and use them here instead of calling this function.

Hellzy committed:
serverless/appsec/config: read the env at init for IsStandalone()
